### PR TITLE
Adding entprom  - a prometheus hook

### PR DESCRIPTION
**Pull request by yonidavidson**
Adds a hook with Readme and examples for using Prometheus hook when mutating data on the graph.

### CODE REVIEW

File: entprom/README.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+Attach hook only for specific schema - for example User
+
+```go
+// Hooks of the User.
+func (User) Hooks() []ent.Hook {
+    return []ent.Hook{
+        entprom.Hook(),
+    }
+}
```
Comment (rotemtam) [inline entprom/README.md:66]: @yonidavidson, will this work if you rn the `entprom` hook on multiple, different schemas?

File: entprom/README.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+Next, wherever you are setting up your HTTP listener:
+```go
+//get from github.com/prometheus/client_golang/prometheus/promhttp
+http.Handle("/metrics", promhttp.Handler())
+```
+
+After your server is running and a few mutations were made, you can get the metrics from the `/metrics` endpoint.
+
+TODO(yoni): add curl example and sample output.
```
Comment (rotemtam) [inline entprom/README.md:34]: please

File: entprom/go.mod [hunk 3]
```diff
@@ -0,0 +1,10 @@
+module entprom
```
Comment (rotemtam) [inline entprom/go.mod:1]: remove this file, use the global one with everyone else. we will split the modules if needed at a later point in time

File: entprom/README.md (hunk 1 above)
Comment (yonidavidson) [reply entprom/README.md]: only if there is a difference in one of the labels.

**Comment by giautm** [comment]
Why not opentelemetry?

**Comment by yonidavidson** [comment]
> Why not opentelemetry?

Ideally we would want to integrate with the OpenTelemetry project, but it appears that the go metrics integration is on hold there:
[1]: The development of the metrics API and SDK has paused due to limited development resources, prioritization of a stable Traces release, and instability of the official overall design from the OpenTelemetry specification. Pull Requests for metrics-related issues are not being accepted currently outside of security vulnerability mitigations.

**Comment by rotemtam** [comment]
I agree with @yonidavidson here. It appears that OpenTelemetry _should_ be the future, but for now, it's on hold and we'd rather have ent users something that's stable.

**Comment by giautm** [comment]
I knew OpenTelemetry was not complete yet, but don't know they pause development on it, sad.

But, [opencensus](https://opencensus.io/stats/) is still here for production usage. Prometheus looks like vendor-lock, If I want to use this hook on GCP/AWS then I have to re-write.

PS: I still usage prometheus on our system, but with opencensus integration.

**Comment by cliedeman** [comment]
Would it be feasible to create a entmetrics package that has different templates for each implementation. I know tracing changes quite a bit for opentelemetry but with metrics it should be a lot more straightforward?

**Comment by yonidavidson** [comment]
> Would it be feasible to create an entmetrics package that has different templates for each implementation. I know tracing changes quite a bit for opentelemetry but with metrics it should be a lot more straightforward?

It's an interesting  idea,  If you can write a few words about it in a new issue I'll be able to review.

**Comment by rotemtam** [comment]
hey @cliedeman @giautm 

thanks for the great feedback. I completely agree that ent should provide a robust solution that is decoupled from any specific metric collection backend. Our users are most likely using many different metric collection backends such as Prometheus, InfluxDB, CloudWatch, datadog, etc. 

This issue is very common and widespread and has spawned initiatives such as [OpenCensus](https://opencensus.io/) and [OpenTelemetry](https://opentelemetry.io/). I am confident that this problem should and will be solved in a standards-based layer where the entire industry will collaborate to decouple the interface and implementation of such systems. Similar things are happening with the [Open Container Initiative](https://opencontainers.org/), [Service Mesh Initiative](https://smi-spec.io/).  Currently, it appears that the industry will converge around [OpenTelemetry](https://opentelemetry.io/) but as @yonidavidson mentioned, support for metrics+go there seems to have been put in the icebox. 

I have big hopes for the standard-based approach and I am following the development of OpenTelemetry. As such, I don't think that `ent` should develop an "in-house" abstraction for metric backends. 

The way I think about this PR is that it offers support for a basic, but well-defined use case, to get something working "out the door" for feedback.  Just as we released `entproto` in a fairly limited scope and iterated over a few months to bring it to the point where it's actually usable, I think this PR is just the first in an initiative to bring proper instrumentation to Ent. 

Consider, for example, that we don't have support for "interceptor" like API yet, which would be required to properly instrument reads from the graph (Get/Update). In other words, this is a _very_ initial step towards something which offers broad support for multiple metric collection backends. 

If we don't see any progress in OpenTelemetry within the next few months towards making metrics generally available in Go, we should definitely consider going with @cliedeman's approach to building some local interface that abstracts the metric collection backend. This interface would have a small subset of the OpenTelemetry API surface limited to the things ent needs. 

I hope you agree that releasing `entprom` is a step in the right direction

**Comment by yonidavidson** [comment]
Closing for discussion.

**Pull request closed without merging**